examples/graphics/shader-graph-chunks-builder.html

```diff
@@ -0,0 +1,176 @@
+    <script src="../../build/playcanvas-shader-nodes.js"></script>
+    <style>
+        body { 
+            margin: 0;
+            overflow-y: hidden;
+        }
+    </style>
+</head>
+
+<body>
+    <!-- The canvas element -->
+    <canvas id="application-canvas"></canvas>
+    
+    <!-- The script -->
+    <script>
+        var entity, light, camera;
+        var time = 0;
+
+        var canvas = document.getElementById("application-canvas");
+
+        // Create the application and start the update loop
+        var app = new pc.Application(canvas);
+
+        // Set the canvas to fill the window and automatically change resolution to be the same as the canvas size
+        app.setCanvasFillMode(pc.FILLMODE_FILL_WINDOW);
+        app.setCanvasResolution(pc.RESOLUTION_AUTO);
+
+        window.addEventListener("resize", function () {
+            app.resizeCanvas(canvas.width, canvas.height);
+        });
+
+        var miniStats = new pcx.MiniStats(app);
+
+        // register core nodes
+        app.shaderNodes.register(pcsg.shaderNodes);
+
+        // set ambient light
+        app.scene.ambientLight = new pc.Color(0.2, 0.2, 0.2);
+
+        // Create an Entity with a camera component
+        camera = new pc.Entity();
+        camera.addComponent("camera", {
+            clearColor: new pc.Color(0.4, 0.45, 0.5)
+        });
+        camera.translate(0, 7, 25);
+        app.root.addChild(camera);
+
+        // Create an entity with a directional light component
+        var light = new pc.Entity();
+        light.addComponent("light", {
+            type: "directional",
+            color: new pc.Color(1, 0.8, 0.25)
+        });
+        app.root.addChild(light);
+        light.setLocalEulerAngles(15, -20, 0);
+
+        // load in a model
+        var url = "../assets/models/statue.glb";
+        app.assets.loadFromUrl(url, "container", function (err, asset) {
+            entity = new pc.Entity();
+            var modelComponent = entity.addComponent("model", {
+                    type: "asset",
+                    asset: asset.resource.model
+                });
+            app.root.addChild(entity);
+            var model = modelComponent.model;
+
+            // load a texture asset
+            app.assets.loadFromUrl("../assets/textures/clouds.jpg", "texture", function (err, asset) {    
+                    var cloudsMap = asset.resource;            
+                app.assets.loadFromUrl("../assets/textures/particles-numbers.png", "texture", function (err, asset) {    
+                    var numbersMap = asset.resource;            
+
+                    // create a shader graph builder
+                    var builder = new pc.ShaderGraphBuilder(app);
+
+                    // add input parameters
+                    var timeParam = builder.addParam('float', 'uTime', 0.0);
+                    var texParamA = builder.addParam('sampler2D', 'uTexMapA', cloudsMap);
```

Comment:
the value (cloudsMap) is described as optional (and should be optional), but crashes when not specified.

Reply:
hmm - probably best to keep it optional, and have the builder put in a placeholder if not supplied - I'll see if I can add a placeholder (this is what materials do (in the material handler) I think)

Reply:
refactored - Texture parameters are now treated separately - with additional validation and placeholder.